### PR TITLE
[SPARK-29651][SQL] Fix parsing of interval seconds fraction

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to extract parsing of the seconds interval units to the private method `parseNanos` in `IntervalUtils` and modify the code to correctly parse the fractional part of the seconds unit of intervals in the cases:
- When the fractional part has less than 9 digits
- The seconds unit is negative

### Why are the changes needed?
The changes are needed to fix the issues:
```sql
spark-sql> select interval '10.123456 seconds';
interval 10 seconds 123 microseconds
```
The correct result must be `interval 10 seconds 123 milliseconds 456 microseconds`
```sql
spark-sql> select interval '-10.123456789 seconds';
interval -9 seconds -876 milliseconds -544 microseconds
```
but the whole interval should be negated, and the result must be `interval -10 seconds -123 milliseconds -456 microseconds`, taking into account the truncation to microseconds.

### Does this PR introduce any user-facing change?
Yes. After changes:
```sql
spark-sql> select interval '10.123456 seconds';
interval 10 seconds 123 milliseconds 456 microseconds
spark-sql> select interval '-10.123456789 seconds';
interval -10 seconds -123 milliseconds -456 microseconds
```

### How was this patch tested?
By existing and new tests in `ExpressionParserSuite`.

### CODE REVIEW

**Comment by yaooqinn** [comment]
This might be fixed in https://github.com/apache/spark/pull/26314, would you please take a look?

**Comment by MaxGekk** [comment]
Probably, this could be fixed by #26314 but tests from this PR are valuable as well, I think. @cloud-fan @dongjoon-hyun WDYT?

**Comment by MaxGekk** [comment]
The bugs exist in version since 2.0, and the fixes can be ported back but the feature in #26314 cannot, I think.

**Comment by dongjoon-hyun** [comment]
I agree with @MaxGekk .

**Comment by cloud-fan** [comment]
not related to this PR, but I find `10 seconds 123 milliseconds 456 microseconds` is hard to read. Shall we only display a fraction value of second?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 1]
```diff
@@ -292,6 +287,13 @@ object IntervalUtils {
     new CalendarInterval(months, microseconds)
   }
 
+  private def parseNanos(nanosStr: String, isNegative: Boolean): Long = {
+    val alignedStr = if (nanosStr == null) nanosStr else (nanosStr + "000000000").substring(0, 9)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: shall we return 0 immediately if `nanosStr` is null? It's easier to read than relying on `toLongWithRange` to return 0.

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
The failure https://github.com/apache/spark/pull/26313#issuecomment-548070328 is not related to the changes, it seems

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by dongjoon-hyun** [comment]
Yes. +1 for @cloud-fan 's suggestion. Originally, I thought it's natural, but @cloud-fan 's suggestion sounds more practical.
> not related to this PR, but I find 10 seconds 123 milliseconds 456 microseconds is hard to read. Shall we only display a fraction value of second?

**Comment by MaxGekk** [comment]
> Yes. +1 for @cloud-fan 's suggestion. Originally, I thought it's natural, but @cloud-fan 's suggestion sounds more practical.

Let's discuss this there: https://issues.apache.org/jira/browse/SPARK-29671

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by MaxGekk** [comment]
Should I back port this on 2.4?

**Comment by cloud-fan** [comment]
yes, please send a backport PR, thanks!

**Comment by MaxGekk** [comment]
Here is the back port to 2.4 https://github.com/apache/spark/pull/26355

**Pull request closed without merging**